### ISSUE TITLE
Compressing assets is not working as expected.

### ISSUE THREAD

**Issue by Locustv2** [yiisoft/yii2#10151]
I am using the extension scroll pager: 

`"require": {
        "kop/yii2-scroll-pager": "dev-master",
    },`

I created an asset bundle to compress the asset *kop\y2sp\assets\InfiniteAjaxScrollAsset* and i have only 1 target 'all' in my compress script.

`
'all' => [
            'class' => 'yii\web\AssetBundle',
            'basePath' => '@webroot',
            'baseUrl' => '@web',
            'js' => 'min/js/all{hash}.js',
            'css' => 'min/css/all{hash}.css',
        ],
`

When i run my application, the compressed js is included but the individual js present in the asset are also included.

FYI: This seems to be happening for widgets only. I tried to compress 2 of kartik assets into 1 file and it is including both the compressed and the original files.

**Comment by Locustv2** [comment]
I think i found the problem.

The original asset is as follows: 

```php
class InfiniteAjaxScrollAsset extends AssetBundle
{
    /**
     * @var array List of bundle class names that this bundle depends on.
     */
    public $depends = [
        'yii\web\JqueryAsset',
    ];

    /**
     * @inheritdoc
     */
    public function init()
    {
        $this->sourcePath = '@vendor/webcreate/jquery-ias/src';
        $this->js = [
            'callbacks.js',
            'jquery-ias.js',
            'extension/history.js',
            'extension/noneleft.js',
            'extension/paging.js',
            'extension/spinner.js',
            'extension/trigger.js'
        ];
    }
}
```

With this type of init(), the compressed asset are being added as well as the original js files.

But if i change the file to this: 

```php
class InfiniteAjaxScrollAsset extends AssetBundle
{
    /**
     * @var array List of bundle class names that this bundle depends on.
     */
    public $depends = [
        'yii\web\JqueryAsset',
    ];

    public $sourcePath = '@vendor/webcreate/jquery-ias/src';

    public $js = [
        'callbacks.js',
        'jquery-ias.js',
        'extension/history.js',
        'extension/noneleft.js',
        'extension/paging.js',
        'extension/spinner.js',
        'extension/trigger.js'
    ];
}
```

It works fine, i.e. only the compressed js file is included.

**Comment by pana1990** [comment]
you should report this issue in repository's extension

**Comment by Locustv2** [comment]
Isn't it something related to yii?

I mean the way js/css files are being set are different but i think they should work the same.
so even if i use the first or second method as above, both should work the same. I think this is yii related rather than extention related.

Let's see what the yii team say about this and then i will report to extention repo if needed.

**Comment by pana1990** [comment]
in basic template it works for me. I think that's problem is extension....

**Comment by Locustv2** [comment]
Did you use the scroll pager asset?

**Comment by pana1990** [comment]
No, i use the following : 

```php
class AppAsset extends AssetBundle
{
    public $depends = [
        'yii\web\YiiAsset',
        'yii\bootstrap\BootstrapAsset',
    ];

    public function init()
    {
          $this->basePath = '@webroot';
          $this->baseUrl = '@web';
          $this->css = ['css/site.css'];
    }
}
```

**Comment by Locustv2** [comment]
Then did you compress your css and js files?

**Comment by pana1990** [comment]
@Locustv2 init method say "If you override this method, make sure you call the parent implementation in the last."

can you try to call parent last method init? i think error is there.

**Comment by Locustv2** [comment]
Already tried, and didn't change anything.

**Comment by pana1990** [comment]
ok, it's truth. Asset should be use properties directly to initialize. I do not think it's a bug if not misuse the init method, imo

**Comment by Clyff** [comment]
I'm facing the same problem here. I have a assetBundle from a package and this asset uses the `init()` method to define the `css`, `js ` and `sourcePath`.

Just in case, I will open an issue in their repo, but i'm wondering if we have any word from the dev team about that. Using `init()` to define properties of an AssetBundle is a bad practice? Or is there any solution to not generate both files?

**Comment by Locustv2** [comment]
Haven't got any replies from the dev till now.

**Comment by dynasource** [comment]
thats because its quite unclear what your issue is. I dont see any issues, except that you are overriding the ancestor init() function without replacing its implementation.

**Comment by Clyff** [comment]
> thats because its quite unclear what your issue is. I dont see any issues, except that you are overriding the ancestor init() function without replacing its implementation.

I will try to make this more clear as possible. This bellow code here:

```php
class MyAsset extends AssetBundle
{
    public $depends = [
        'other/AssetBundle',
    ];

    public function init()
    {
        $this->sourcePath ="some/path";
        $this->css = [
            'style.css'
        ];
        $this->js = [
            'script.js'
        ];
        parent::init();
    }
}
```

Is a valid Asset for compressing? If so, how, **in the application**, not load the original css and js files? Let me know if I need to add more info.

**Comment by Clyff** [comment]
Here is an step-by-step of what i did to simulate the problem:

- Install Yii2: composer: `create-project --prefer-dist --stability=dev yiisoft/yii2-app-basic basic`
- Inside the application, run `yii asset/template assets.php`
- Configure the generated file as the following (In my case I'm using gulp to compress the files):

```php
<?php
/**
 * Configuration file for the "yii asset" console command.
 */

// In the console environment, some path aliases may not exist. Please define these:
Yii::setAlias('@webroot', __DIR__ . '/web');
Yii::setAlias('@web', '/');

return [
    // Adjust command/callback for JavaScript files compressing:
    'jsCompressor' => 'gulp assetScripts --src {from} --dist {to}',
    // Adjust command/callback for CSS files compressing:
    'cssCompressor' => 'gulp assetStyles --src {from} --dist {to}',
    // The list of asset bundles to compress:
    'bundles' => [
        'app\assets\AppAsset',
        'yii\web\YiiAsset',
        'yii\web\JqueryAsset',
    ],
    // Asset bundle for compression output:
    'targets' => [
        'all' => [
            'class' => 'yii\web\AssetBundle',
            'basePath' => '@webroot/assets',
            'baseUrl' => '@web/assets',
            'js' => 'all.js',
            'css' => 'all.css',
        ],
    ],
    // Asset manager configuration:
    'assetManager' => [
        'basePath' => '@webroot/assets',
        'baseUrl' => '@web/assets',
    ],
];
```

- In config/web.php add the following component:

```php
<?php
$bundles = require(__DIR__ . '/assets-prod.php');

$config = [
    ...
    components => [
        'assetManager' => [
            'bundles' => $bundles,
            'appendTimestamp' => true,
        ],
        ...
```

- Now, run `yii asset assets.php config/assets-prod.php` to compress.

If you check the `assets-prod.php` file, you will see that the compressing went fine and all assets are empty and depending of `all`.

But, if I overwrite the `AppAsset` to look like this:

```php
<?php
namespace app\assets;

use yii\web\AssetBundle;

class AppAsset extends AssetBundle
{
    public $depends = [
        'yii\web\YiiAsset',
        'yii\bootstrap\BootstrapAsset',
    ];

    public function init()
    {
        $this->basePath = '@webroot';
        $this->baseUrl = '@web';
        $this->css = ['css/site.css'];

        parent::init();
    }
}
```

The compressing will still work, however in my application I can see the original `site.css` being loaded:

![compress](https://cloud.githubusercontent.com/assets/7630386/15858417/bc541494-2c96-11e6-8bc4-090c533a6c5e.jpg)

So, my question is: Is this the expected behavior of the Asset? Shouldn't, after compressed, the asset never being able to rewrite their files somehow?

Or you guys think that using the `init()` method to add properties is a bad practice for this class? As i said, I'm using an extension that have this situation and I already opened an issue there. I just would like to hear what you guys think about it or if i did something wrong in the process described.

**Comment by schmunk42** [comment]
I can confirm the issue like described by @Clyff 

When you try to bundle assets, which change their properties in their `init()` method - you're very likely about to run into problems, like duplicated (but missing) imports, wrong paths, etc....

> Or you guys think that using the init() method to add properties is a bad practice for this class?

I'd say yes.

CC: @Quexer69 @tonydspaniard

**Comment by klimov-paul** [comment]
Relates #12017

**Comment by klimov-paul** [comment]
Resolved by commit 65c60a0c2477117d07a64f45af8fa6c58f3edf62
